Source/ACE/Entity/Vendor.cs

```diff
@@ -58,34 +36,139 @@ public override void OnUse(ObjectGuid playerId)
                     return;
                 }
 
-                string serverMessage = null;
-                // validate within use range, taking into account the radius of the model itself, as well
-                SetupModel csetup = SetupModel.ReadFromDat(PhysicsData.CSetup);
-                float radiusSquared = (GameData.UseRadius + csetup.Radius) * (GameData.UseRadius + csetup.Radius);
+                if (!player.IsWithinUseRadiusOf(this))
+                    player.DoMoveTo(this);
+                else
+                {
+                    chain.AddAction(this, () =>
+                    {
+                        LoadInventory();
+                        UseVendor(player);
+                    });          
+                }
+            });
+
+            chain.EnqueueChain();
+        }
 
-                // We're static, so this is safe -- our Location is never written
-                if (player.Location.SquaredDistanceTo(Location) >= radiusSquared)
+        /// <summary>
+        /// Fired when user approaches vendor, sends all items vendor has for sale to players
+        /// local tracked item list and opens vendor inventory screen. 
+        /// </summary>
+        /// <param name="player"></param>
+        private void UseVendor(Player player)
+        {
+            foreach (KeyValuePair<ObjectGuid, WorldObject> wo in defaultItemsForSale)
+            {
+                player.TrackInteractiveObject(wo.Value);
```

Comment:
This is called from an ActionChain on the vendor, but it modifies the player, this cannot happen.

Instead you should make a list of available items from the vendor, then create an actionchain on the player passing that list.